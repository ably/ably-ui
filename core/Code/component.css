@layer base {
  @import url("https://fonts.googleapis.com/css2?family=Source+Code+Pro:wght@600&display=swap");
}

@layer components {
  .hljs {
    background: var(--syntax-black);
    color: var(--syntax-light-grey);
  }

  .hljs-emphasis {
    font-style: italic;
  }

  .hljs-strong {
    font-weight: bold;
  }

  .hljs-comment,
  .hljs-quote {
    color: var(--syntax-dark-grey);
  }

  .hljs-variable,
  .hljs-template-variable,
  .hljs-tag,
  .hljs-name,
  .hljs-selector-id,
  .hljs-selector-class,
  .hljs-regexp,
  .hljs-deletion {
    color: var(--syntax-red);
  }

  .hljs-number,
  .hljs-built_in,
  .hljs-literal,
  .hljs-type,
  .hljs-params,
  .hljs-meta,
  .hljs-link {
    color: var(--syntax-orange);
  }

  .hljs-attribute {
    color: var(--syntax-yellow);
  }

  .hljs-string,
  .hljs-symbol,
  .hljs-bullet,
  .hljs-addition {
    color: var(--syntax-green);
  }

  .hljs-title,
  .hljs-section {
    color: var(--syntax-blue);
  }

  .hljs-keyword,
  .hljs-selector-tag {
    color: var(--syntax-lilac);
  }

  .hljs-subst {
    color: var(--syntax-mid-grey);
  }
}


/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIndlYnBhY2s6Ly9BYmx5VWkuQ29yZS5bbmFtZV0vLi9zcmMvY29yZS9Db2RlL2NvbXBvbmVudC5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSw2RkFBNkY7QUFDL0Y7O0FBRUE7RUFDRTtJQUNFLCtCQUErQjtJQUMvQiwrQkFBK0I7RUFDakM7O0VBRUE7SUFDRSxrQkFBa0I7RUFDcEI7O0VBRUE7SUFDRSxpQkFBaUI7RUFDbkI7O0VBRUE7O0lBRUUsOEJBQThCO0VBQ2hDOztFQUVBOzs7Ozs7OztJQVFFLHdCQUF3QjtFQUMxQjs7RUFFQTs7Ozs7OztJQU9FLDJCQUEyQjtFQUM3Qjs7RUFFQTtJQUNFLDJCQUEyQjtFQUM3Qjs7RUFFQTs7OztJQUlFLDBCQUEwQjtFQUM1Qjs7RUFFQTs7SUFFRSx5QkFBeUI7RUFDM0I7O0VBRUE7O0lBRUUsMEJBQTBCO0VBQzVCOztFQUVBO0lBQ0UsNkJBQTZCO0VBQy9CO0FBQ0YiLCJmaWxlIjoiY29yZS9Db2RlL2NvbXBvbmVudC5jc3MiLCJzb3VyY2VzQ29udGVudCI6WyJAbGF5ZXIgYmFzZSB7XG4gIEBpbXBvcnQgdXJsKFwiaHR0cHM6Ly9mb250cy5nb29nbGVhcGlzLmNvbS9jc3MyP2ZhbWlseT1Tb3VyY2UrQ29kZStQcm86d2dodEA2MDAmZGlzcGxheT1zd2FwXCIpO1xufVxuXG5AbGF5ZXIgY29tcG9uZW50cyB7XG4gIC5obGpzIHtcbiAgICBiYWNrZ3JvdW5kOiB2YXIoLS1zeW50YXgtYmxhY2spO1xuICAgIGNvbG9yOiB2YXIoLS1zeW50YXgtbGlnaHQtZ3JleSk7XG4gIH1cblxuICAuaGxqcy1lbXBoYXNpcyB7XG4gICAgZm9udC1zdHlsZTogaXRhbGljO1xuICB9XG5cbiAgLmhsanMtc3Ryb25nIHtcbiAgICBmb250LXdlaWdodDogYm9sZDtcbiAgfVxuXG4gIC5obGpzLWNvbW1lbnQsXG4gIC5obGpzLXF1b3RlIHtcbiAgICBjb2xvcjogdmFyKC0tc3ludGF4LWRhcmstZ3JleSk7XG4gIH1cblxuICAuaGxqcy12YXJpYWJsZSxcbiAgLmhsanMtdGVtcGxhdGUtdmFyaWFibGUsXG4gIC5obGpzLXRhZyxcbiAgLmhsanMtbmFtZSxcbiAgLmhsanMtc2VsZWN0b3ItaWQsXG4gIC5obGpzLXNlbGVjdG9yLWNsYXNzLFxuICAuaGxqcy1yZWdleHAsXG4gIC5obGpzLWRlbGV0aW9uIHtcbiAgICBjb2xvcjogdmFyKC0tc3ludGF4LXJlZCk7XG4gIH1cblxuICAuaGxqcy1udW1iZXIsXG4gIC5obGpzLWJ1aWx0X2luLFxuICAuaGxqcy1saXRlcmFsLFxuICAuaGxqcy10eXBlLFxuICAuaGxqcy1wYXJhbXMsXG4gIC5obGpzLW1ldGEsXG4gIC5obGpzLWxpbmsge1xuICAgIGNvbG9yOiB2YXIoLS1zeW50YXgtb3JhbmdlKTtcbiAgfVxuXG4gIC5obGpzLWF0dHJpYnV0ZSB7XG4gICAgY29sb3I6IHZhcigtLXN5bnRheC15ZWxsb3cpO1xuICB9XG5cbiAgLmhsanMtc3RyaW5nLFxuICAuaGxqcy1zeW1ib2wsXG4gIC5obGpzLWJ1bGxldCxcbiAgLmhsanMtYWRkaXRpb24ge1xuICAgIGNvbG9yOiB2YXIoLS1zeW50YXgtZ3JlZW4pO1xuICB9XG5cbiAgLmhsanMtdGl0bGUsXG4gIC5obGpzLXNlY3Rpb24ge1xuICAgIGNvbG9yOiB2YXIoLS1zeW50YXgtYmx1ZSk7XG4gIH1cblxuICAuaGxqcy1rZXl3b3JkLFxuICAuaGxqcy1zZWxlY3Rvci10YWcge1xuICAgIGNvbG9yOiB2YXIoLS1zeW50YXgtbGlsYWMpO1xuICB9XG5cbiAgLmhsanMtc3Vic3Qge1xuICAgIGNvbG9yOiB2YXIoLS1zeW50YXgtbWlkLWdyZXkpO1xuICB9XG59XG4iXSwic291cmNlUm9vdCI6IiJ9*/