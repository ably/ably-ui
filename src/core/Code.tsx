import React from "react";
import {
  highlightSnippet,
  registerDefaultLanguages,
} from "./utils/syntax-highlighter";
import languagesRegistry from "./utils/syntax-highlighter-registry";
import cn from "./utils/cn";

registerDefaultLanguages(languagesRegistry);

type CodeProps = {
  language: string;
  snippet: string;
  textSize?: string;
  padding?: string;
  additionalCSS?: string;
  showLines?: boolean;
  lineCSS?: string;
  wrap?: boolean;
};

const Code = ({
  language,
  snippet,
  textSize = "ui-text-code",
  padding = "p-8",
  additionalCSS = "",
  showLines,
  lineCSS,
  wrap = false,
}: CodeProps) => {
  // Trim the snippet and remove trailing empty lines
  const trimmedSnippet = snippet.trimEnd();
  const HTMLraw = highlightSnippet(language, trimmedSnippet) ?? "";
  const className = `language-${language} ${textSize}`;

  // Calculate line count after removing trailing empty lines
  const lines = trimmedSnippet.split(/\r\n|\r|\n/);
  const lineCount = lines.length;

  return (
    <div
      className={cn("hljs overflow-y-auto flex", padding, additionalCSS)}
      data-id="code"
    >
      {showLines ? (
        <div className="text-p4 leading-normal pt-px">
          {[...Array(lineCount)].map((_, i) => (
            <p
              className={cn(
                "mr-4 font-mono text-right text-neutral-800",
                lineCSS,
              )}
              key={i}
            >
              {i + 1}
            </p>
          ))}
        </div>
      ) : null}
      <pre
        lang={language}
        className={cn(
          "h-full flex-1 text-p4 leading-normal",
          wrap ? "whitespace-pre-wrap break-words" : "overflow-x-auto",
        )}
      >
        <code
          className={className}
          dangerouslySetInnerHTML={{ __html: HTMLraw }}
        />
      </pre>
    </div>
  );
};

export default Code;
