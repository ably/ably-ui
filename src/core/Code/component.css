@import "../utils/syntax-highlighter.css";
