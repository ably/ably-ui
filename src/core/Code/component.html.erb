<div class="hljs p-32 overflow-auto" data-id="code">
  <pre lang="<%= @language %>"><code class="font-mono language-<%= @language %>"><%= @snippet %></code></pre>
</div>
