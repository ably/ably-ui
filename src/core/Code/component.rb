module AblyUi
  module Core
    class Code < ViewComponent::Base
      include AblyUi::Core::Util
      def initialize(language:, snippet:, text_size: 'ui-text-code')
        @language = language
        @snippet = snippet.try(:strip)
        @text_size = text_size
      end
    end
  end
end
