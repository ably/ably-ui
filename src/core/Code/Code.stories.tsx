import Code from "../Code";

export default {
  title: "Components/Code",
  component: Code,
  tags: ["autodocs"],
};

export const Javascript = {
  args: {
    language: "javascript",
    snippet: `var ably = new Ably.Realtime('1WChTA.mc0Biw:kNfiYG4KiPgmHHgH');
var channel = ably.channels.get('web-pal');
    
// Subscribe to messages on channel
channel.subscribe('greeting', function(message) {
  alert(message.data);
});`,
    additionalCSS: "bg-neutral-1200 p-16",
  },
};

export const Swift = {
  args: {
    language: "swift",
    snippet: `let ably = ARTRealtime(key: "1WChTA.mc0Biw:kNfiYG4KiPgmHHgH")
let channel = ably.channels.get("web-pal")

// Subscribe to messages on channel
channel.subscribe("greeting") { message in
  print("\\(message.data)")
}`,
  },
};

export const Java = {
  args: {
    language: "java",
    snippet: `AblyRealtime ably = new AblyRealtime("1WChTA.mc0Biw:kNfiYG4KiPgmHHgH");
Channel channel = ably.channels.get("web-pal");

/* Subscribe to messages on channel */

MessageListener listener;
listener = new MessageListener() {
  @Override
  public void onMessage(Message message) {
    System.out.print(message.data);
  };
};
channel.subscribe("greeting", listener);`,
  },
};

export const Kotlin = {
  args: {
    language: "kotlin",
    snippet: `var ably = new Ably.Realtime('1WChTA.mc0Biw:kNfiYG4KiPgmHHgH');
val exampleConstraints = DefaultResolutionConstraints(
  DefaultResolutionSet( // this constructor provides one Resolution for all states
      Resolution(
          accuracy = Accuracy.BALANCED,
          desiredInterval = 1000L,
          minimumDisplacement = 1.0
      )
  ),
  proximityThreshold = DefaultProximity(spatial = 1.0),
  batteryLevelThreshold = 10.0f,
  lowBatteryMultiplier = 2.0f
)`,
  },
};

export const CodeWithLines = {
  args: {
    showLines: true,
    language: "javascript",
    snippet: `var ably = new Ably.Realtime('1WChTA.mc0Biw:kNfiYG4KiPgmHHgH');
var channel = ably.channels.get('web-pal');
    
// Subscribe to messages on channel
channel.subscribe('greeting', function(message) {
  alert(message.data);
});`,
  },
};
