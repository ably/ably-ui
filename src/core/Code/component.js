import "./component.css";
import hljs from "highlight.js/lib/core";

import bash from "highlight.js/lib/languages/bash";
import cpp from "highlight.js/lib/languages/cpp";
import csharp from "highlight.js/lib/languages/csharp";
import dart from "highlight.js/lib/languages/dart";
import plaintext from "highlight.js/lib/languages/plaintext";
import go from "highlight.js/lib/languages/go";
import java from "highlight.js/lib/languages/java";
import javascript from "highlight.js/lib/languages/javascript";
import json from "highlight.js/lib/languages/json";
import objectivec from "highlight.js/lib/languages/objectivec";
import php from "highlight.js/lib/languages/php";
import python from "highlight.js/lib/languages/python";
import ruby from "highlight.js/lib/languages/ruby";
import swift from "highlight.js/lib/languages/swift";
import xml from "highlight.js/lib/languages/xml";

// Supported languages need to be imported here
// https://github.com/highlightjs/highlight.js/blob/master/SUPPORTED_LANGUAGES.md
const register = [
  { label: "", key: "plaintext", module: plaintext },
  { label: "JS", key: "javascript", module: javascript },
  { label: "Java", key: "java", module: java },
  { label: "Ruby", key: "ruby", module: ruby },
  { label: "Python", key: "python", module: python },
  { label: "PHP", key: "php", module: php },
  { label: "Shell", key: "bash", module: bash },
  { label: "C#", key: "cs", module: csharp },
  { label: "Go", key: "go", module: go },
  { label: "HTML", key: "xml", module: xml },
  { label: "C++", key: "cpp", module: cpp },
  { label: "Dart", key: "dart", module: dart },
  { label: "Swift", key: "swift", module: swift },
  { label: "Objective C", key: "objectivec", module: objectivec },
  { label: "Node.js", key: "javascript", module: javascript },
  { label: "JSON", key: "json", module: json },
];

// Initialize a sub-set of languages as used
register.forEach(({ key, module }) => hljs.registerLanguage(key, module));

function convertCodeToHTML(el) {
  if (!el) throw "Missing code element";

  const pre = el.querySelector("pre");
  const code = el.querySelector("code");
  const language = pre.getAttribute("lang");

  if (!code || !pre || !language) throw "Malformed code element";

  const { innerHTML } = code;
  const html = hljs.highlight(innerHTML, { language }).value;

  code.innerHTML = html;
}

function highlight(language, snippet) {
  return hljs.highlight(snippet, { language }).value;
}

export { highlight };
export default convertCodeToHTML;
