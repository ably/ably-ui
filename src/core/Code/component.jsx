import React from "react";
import T from "prop-types";

import "./component.css";
import { highlight } from "./component.js";

const Code = ({ language, snippet, textSize = "ui-text-code", padding = "p-32", additionalCSS = "" }) => {
  const HTMLraw = highlight(language, `${snippet}`.trim());
  const className = `language-${language} ${textSize}`;
  return (
    <div className={`hljs overflow-auto ${padding} ${additionalCSS}`} data-id="code">
      <pre lang={language}>
        <code className={className} dangerouslySetInnerHTML={{ __html: HTMLraw }} />
      </pre>
    </div>
  );
};

Code.propTypes = {
  language: T.string,
  snippet: T.string,
  textSize: T.string,
  padding: T.string,
  additionalCSS: T.string,
};

export default Code;
