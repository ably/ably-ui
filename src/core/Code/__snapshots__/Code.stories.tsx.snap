// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS Components/Code CodeWithLines smoke-test 1`] = `
<div class="hljs overflow-auto flex p-32 "
     data-id="code"
>
  <div>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      1
    </p>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      2
    </p>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      3
    </p>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      4
    </p>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      5
    </p>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      6
    </p>
    <p class="mr-24 font-mono text-right text-neutral-800 ">
      7
    </p>
  </div>
  <pre lang="javascript">
    <code class="language-javascript ui-text-code">
      <span class="hljs-keyword">
        var
      </span>
      ably =
      <span class="hljs-keyword">
        new
      </span>
      <span class="hljs-title class_">
        Ably
      </span>
      .
      <span class="hljs-title class_">
        Realtime
      </span>
      (
      <span class="hljs-string">
        '1WChTA.mc0Biw:kNfiYG4KiPgmHHgH'
      </span>
      );
      <span class="hljs-keyword">
        var
      </span>
      channel = ably.
      <span class="hljs-property">
        channels
      </span>
      .
      <span class="hljs-title function_">
        get
      </span>
      (
      <span class="hljs-string">
        'web-pal'
      </span>
      );
      <span class="hljs-comment">
        // Subscribe to messages on channel
      </span>
      channel.
      <span class="hljs-title function_">
        subscribe
      </span>
      (
      <span class="hljs-string">
        'greeting'
      </span>
      ,
      <span class="hljs-keyword">
        function
      </span>
      (
      <span class="hljs-params">
        message
      </span>
      ) {
      <span class="hljs-title function_">
        alert
      </span>
      (message.
      <span class="hljs-property">
        data
      </span>
      );
});
    </code>
  </pre>
</div>
`;

exports[`JS Components/Code Java smoke-test 1`] = `
<div class="hljs overflow-auto flex p-32 "
     data-id="code"
>
  <pre lang="java">
    <code class="language-java ui-text-code">
      <span class="hljs-type">
        AblyRealtime
      </span>
      <span class="hljs-variable">
        ably
      </span>
      <span class="hljs-operator">
        =
      </span>
      <span class="hljs-keyword">
        new
      </span>
      <span class="hljs-title class_">
        AblyRealtime
      </span>
      (
      <span class="hljs-string">
        "1WChTA.mc0Biw:kNfiYG4KiPgmHHgH"
      </span>
      );
      <span class="hljs-type">
        Channel
      </span>
      <span class="hljs-variable">
        channel
      </span>
      <span class="hljs-operator">
        =
      </span>
      ably.channels.get(
      <span class="hljs-string">
        "web-pal"
      </span>
      );
      <span class="hljs-comment">
        /* Subscribe to messages on channel */
      </span>
      MessageListener listener;
listener =
      <span class="hljs-keyword">
        new
      </span>
      <span class="hljs-title class_">
        MessageListener
      </span>
      () {
      <span class="hljs-meta">
        @Override
      </span>
      <span class="hljs-keyword">
        public
      </span>
      <span class="hljs-keyword">
        void
      </span>
      <span class="hljs-title function_">
        onMessage
      </span>
      <span class="hljs-params">
        (Message message)
      </span>
      {
    System.out.print(message.data);
  };
};
channel.subscribe(
      <span class="hljs-string">
        "greeting"
      </span>
      , listener);
    </code>
  </pre>
</div>
`;

exports[`JS Components/Code Javascript smoke-test 1`] = `
<div class="hljs overflow-auto flex p-32 "
     data-id="code"
>
  <pre lang="javascript">
    <code class="language-javascript ui-text-code">
      <span class="hljs-keyword">
        var
      </span>
      ably =
      <span class="hljs-keyword">
        new
      </span>
      <span class="hljs-title class_">
        Ably
      </span>
      .
      <span class="hljs-title class_">
        Realtime
      </span>
      (
      <span class="hljs-string">
        '1WChTA.mc0Biw:kNfiYG4KiPgmHHgH'
      </span>
      );
      <span class="hljs-keyword">
        var
      </span>
      channel = ably.
      <span class="hljs-property">
        channels
      </span>
      .
      <span class="hljs-title function_">
        get
      </span>
      (
      <span class="hljs-string">
        'web-pal'
      </span>
      );
      <span class="hljs-comment">
        // Subscribe to messages on channel
      </span>
      channel.
      <span class="hljs-title function_">
        subscribe
      </span>
      (
      <span class="hljs-string">
        'greeting'
      </span>
      ,
      <span class="hljs-keyword">
        function
      </span>
      (
      <span class="hljs-params">
        message
      </span>
      ) {
      <span class="hljs-title function_">
        alert
      </span>
      (message.
      <span class="hljs-property">
        data
      </span>
      );
});
    </code>
  </pre>
</div>
`;

exports[`JS Components/Code Kotlin smoke-test 1`] = `
<div class="hljs overflow-auto flex p-32 "
     data-id="code"
>
  <pre lang="kotlin">
    <code class="language-kotlin ui-text-code">
      <span class="hljs-keyword">
        var
      </span>
      ably = new Ably.Realtime(
      <span class="hljs-string">
        '1WChTA.mc0Biw:kNfiYG4KiPgmHHgH'
      </span>
      );
      <span class="hljs-keyword">
        val
      </span>
      exampleConstraints = DefaultResolutionConstraints(
  DefaultResolutionSet(
      <span class="hljs-comment">
        // this constructor provides one Resolution for all states
      </span>
      Resolution(
          accuracy = Accuracy.BALANCED,
          desiredInterval =
      <span class="hljs-number">
        1000L
      </span>
      ,
          minimumDisplacement =
      <span class="hljs-number">
        1.0
      </span>
      )
  ),
  proximityThreshold = DefaultProximity(spatial =
      <span class="hljs-number">
        1.0
      </span>
      ),
  batteryLevelThreshold =
      <span class="hljs-number">
        10.0f
      </span>
      ,
  lowBatteryMultiplier =
      <span class="hljs-number">
        2.0f
      </span>
      )
    </code>
  </pre>
</div>
`;

exports[`JS Components/Code Swift smoke-test 1`] = `
<div class="hljs overflow-auto flex p-32 "
     data-id="code"
>
  <pre lang="swift">
    <code class="language-swift ui-text-code">
      <span class="hljs-keyword">
        let
      </span>
      ably
      <span class="hljs-operator">
        =
      </span>
      <span class="hljs-type">
        ARTRealtime
      </span>
      (key:
      <span class="hljs-string">
        "1WChTA.mc0Biw:kNfiYG4KiPgmHHgH"
      </span>
      )
      <span class="hljs-keyword">
        let
      </span>
      channel
      <span class="hljs-operator">
        =
      </span>
      ably.channels.get(
      <span class="hljs-string">
        "web-pal"
      </span>
      )
      <span class="hljs-comment">
        // Subscribe to messages on channel
      </span>
      channel.subscribe(
      <span class="hljs-string">
        "greeting"
      </span>
      ) { message
      <span class="hljs-keyword">
        in
      </span>
      <span class="hljs-built_in">
        print
      </span>
      (
      <span class="hljs-string">
        "
        <span class="hljs-subst">
          \\(message.data)
        </span>
        "
      </span>
      )
}
    </code>
  </pre>
</div>
`;
